ClassCastException in HandlerFactories

There are reports from a number of bindings about ClassCastExceptions, when trying to instantiate a handler for a bridge - the Thing framework seems to hand in a Thing (and not a Bridge) instance in that case. Before the Thing refactoring, this code was working, so it seems that this must be some bug in the framework, not in the bindings. See a log here:

```
Caused by: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
at org.openhab.binding.homematic.handler.HomematicThingHandlerFactory.createHandler(HomematicThingHandlerFactory.java:50)[204:org.openhab.binding.homematic:2.0.0.201611212345]
at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:98)[104:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
at org.eclipse.smarthome.core.thing.internal.ThingManager$7.call(ThingManager.java:586)[104:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
at org.eclipse.smarthome.core.thing.internal.ThingManager$7.call(ThingManager.java:1)[104:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:177)[97:org.eclipse.smarthome.core:0.9.0.201611191809]
... 4 more
```

More infos also at https://github.com/openhab/openhab2-addons/issues/1462

i activate the logging in  karaf with these both lines and hope this was correct.
log:set DEBUG org.org.eclipse.smarthome.*
log:set DEBUG org.openhab.binding.homematic.*

Then i uninstall the binding and installed it again. For the new installe here its the logging:

2016-11-23 16:15:19.455 [DEBUG] [org.openhab.binding.homematic       ] - BundleEvent INSTALLED - org.openhab.binding.homematic
2016-11-23 16:15:19.728 [DEBUG] [org.openhab.binding.homematic       ] - BundleEvent RESOLVED - org.openhab.binding.homematic
2016-11-23 16:15:19.756 [DEBUG] [org.openhab.binding.homematic       ] - BundleEvent STARTING - org.openhab.binding.homematic
2016-11-23 16:15:19.759 [DEBUG] [org.openhab.binding.homematic       ] - BundleEvent STARTED - org.openhab.binding.homematic
2016-11-23 16:15:19.844 [DEBUG] [org.openhab.binding.homematic       ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.type.ChannelTypeProvider, 

org.openhab.binding.homematic.type.HomematicChannelTypeProvider}={component.name=org.openhab.binding.homematic.type.HomematicChannelTypeProviderImpl, component.id=189, service.id=317, 

service.bundleid=203, service.scope=bundle} - org.openhab.binding.homematic
2016-11-23 16:15:19.859 [DEBUG] [org.openhab.binding.homematic       ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingTypeProvider, 

org.openhab.binding.homematic.type.HomematicThingTypeProvider}={component.name=org.openhab.binding.homematic.type.HomematicThingTypeProviderImpl, component.id=190, service.id=318, 

service.bundleid=203, service.scope=bundle} - org.openhab.binding.homematic
2016-11-23 16:15:19.879 [DEBUG] [org.openhab.binding.homematic       ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.config.core.ConfigDescriptionProvider, 

org.openhab.binding.homematic.type.HomematicConfigDescriptionProvider}={component.name=org.openhab.binding.homematic.type.HomematicConfigDescriptionProviderImpl, component.id=192, 

service.id=319, service.bundleid=203, service.scope=bundle} - org.openhab.binding.homematic
2016-11-23 16:15:19.889 [DEBUG] [org.openhab.binding.homematic       ] - ServiceEvent REGISTERED - {org.openhab.binding.homematic.type.HomematicTypeGenerator}=

{component.name=org.openhab.binding.homematic.type.HomematicTypeGenerator, component.id=191, service.id=320, service.bundleid=203, service.scope=bundle} - org.openhab.binding.homematic
2016-11-23 16:15:19.911 [DEBUG] [org.openhab.binding.homematic       ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory}=

{component.name=org.openhab.binding.homematic.handler.HomematicThingHandlerFactory, component.id=193, service.id=321, service.bundleid=203, service.scope=bundle} - 

org.openhab.binding.homematic
2016-11-23 16:15:19.946 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while calling thing handler factory 

'org.openhab.binding.homematic.handler.HomematicThingHandlerFactory@85537': java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to 

org.eclipse.smarthome.core.thing.Bridge
java.util.concurrent.ExecutionException: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_111]
        at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_111]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:186)
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:79)
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:63)
        at org.eclipse.smarthome.core.thing.internal.ThingManager.doRegisterHandler(ThingManager.java:582)
        at org.eclipse.smarthome.core.thing.internal.ThingManager.registerHandler(ThingManager.java:561)
        at org.eclipse.smarthome.core.thing.internal.ThingManager.addThingHandlerFactory(ThingManager.java:1024)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)[:1.8.0_111]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)[:1.8.0_111]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)[:1.8.0_111]
        at java.lang.reflect.Method.invoke(Method.java:498)[:1.8.0_111]
        at org.apache.felix.scr.impl.helper.BaseMethod.invokeMethod(BaseMethod.java:222)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.helper.BaseMethod.access$500(BaseMethod.java:37)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.helper.BaseMethod$Resolved.invoke(BaseMethod.java:615)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.helper.BaseMethod.invoke(BaseMethod.java:499)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.helper.BindMethod.invoke(BindMethod.java:41)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.DependencyManager.doInvokeBindMethod(DependencyManager.java:1660)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.DependencyManager.invokeBindMethod(DependencyManager.java:1636)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.invokeBindMethod(SingleComponentManager.java:370)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:319)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:295)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1215)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1136)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.trackAdding(ServiceTracker.java:945)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.track(ServiceTracker.java:881)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:1167)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.BundleComponentActivator$ListenerInfo.serviceChanged(BundleComponentActivator.java:120)[34:org.apache.felix.scr:2.0.2]
        at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:914)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:127)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:225)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:464)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:869)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:857)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:133)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.registerService(AbstractComponentManager.java:915)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:715)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:399)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.config.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:676)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:339)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:360)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.Activator.access$000(Activator.java:53)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:260)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.utils.extender.AbstractExtender.createExtension(AbstractExtender.java:259)[34:org.apache.felix.scr:2.0.2]
        at org.apache.felix.utils.extender.AbstractExtender.modifiedBundle(AbstractExtender.java:232)[34:org.apache.felix.scr:2.0.2]
        at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:482)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:415)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:444)[org.osgi.core-6.0.0.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:902)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:165)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:75)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:67)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:102)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.Module.publishEvent(Module.java:466)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.container.Module.start(Module.java:457)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:393)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:412)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.startBundle(FeaturesServiceImpl.java:1199)[9:org.apache.karaf.features.core:4.0.4]
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:840)[9:org.apache.karaf.features.core:4.0.4]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1089)[9:org.apache.karaf.features.core:4.0.4]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl$1.call(FeaturesServiceImpl.java:985)[9:org.apache.karaf.features.core:4.0.4]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_111]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_111]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_111]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_111]
Caused by: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
        at org.openhab.binding.homematic.handler.HomematicThingHandlerFactory.createHandler(HomematicThingHandlerFactory.java:50)[203:org.openhab.binding.homematic:2.0.0.201611222318]
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:98)[104:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$7.call(ThingManager.java:586)[104:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$7.call(ThingManager.java:1)[104:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
        at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:177)[97:org.eclipse.smarthome.core:0.9.0.201611191809]
        ... 4 more
2016-11-23 16:15:19.982 [DEBUG] [org.openhab.binding.homematic       ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.config.discovery.UpnpDiscoveryParticipant}=

{component.name=org.eclipse.smarthome.binding.homematic.discovery.bridge, component.id=194, service.id=322, service.bundleid=203, service.scope=bundle} - org.openhab.binding.homematic

Just for the record: This happens during the netatmo binding initialization too:

    Caused by: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
            at org.openhab.binding.netatmo.internal.NetatmoHandlerFactory.createHandler(NetatmoHandlerFactory.java:53)[196:org.openhab.binding.netatmo:2.0.0.201611181914]
            at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:98)[103:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
            at org.eclipse.smarthome.core.thing.internal.ThingManager$7.call(ThingManager.java:586)[103:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
            at org.eclipse.smarthome.core.thing.internal.ThingManager$7.call(ThingManager.java:1)[103:org.eclipse.smarthome.core.thing:0.9.0.201611191809]
            at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:177)[96:org.eclipse.smarthome.core:0.9.0.201611191809]
            at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_111]

--- a/bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/BundleProcessorVetoManager.java
+++ b/bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/BundleProcessorVetoManager.java
@@ -1,0 +1,145 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.config.core;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.eclipse.smarthome.config.core.BundleProcessor.BundleProcessorListener;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.LinkedListMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+/**
+ * Utility class in order to delay an action until a binding's XML loading completed.
+ *
+ * It takes an action and whenever it is instructed to execute this action, it is going to check first if there are any
+ * vetoes due to the fact that the XML meta-data of the relevant binding might not be fully loaded. If this is the case,
+ * then the action will be delayed until the loading was finished.
+ *
+ * As this class is not an OSGi service itself, it needs some help from the outside by injecting/removing
+ * BundleProcessors via {@link #addBundleProcessor(BundleProcessor)} and
+ * {@link #removeBundleProcessor(BundleProcessor)}.
+ *
+ * @author Simon Kaufmann - Initial contribution and API
+ */
+public class BundleProcessorVetoManager<T> implements BundleProcessorListener {
+
+    public interface Action<T> {
+        void apply(T object);
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(BundleProcessorVetoManager.class);
+
+    private final Set<BundleProcessor> bundleProcessors = new HashSet<BundleProcessor>();
+    private final Multimap<Long, BundleProcessor> vetoes = Multimaps
+            .synchronizedListMultimap(LinkedListMultimap.<Long, BundleProcessor> create());
+    private final Multimap<Long, T> queue = Multimaps.synchronizedListMultimap(LinkedListMultimap.<Long, T> create());
+    private final Action<T> action;
+
+    /**
+     * Construct a BundleProcessorVetoManager for the given action
+     *
+     * @param action the action to run (potentially delayed)
+     */
+    public BundleProcessorVetoManager(Action<T> action) {
+        this.action = action;
+    }
+
+    @Override
+    public void bundleFinished(final BundleProcessor context, final Bundle bundle) {
+        vetoes.remove(bundle.getBundleId(), context);
+        if (vetoes.get(bundle.getBundleId()).isEmpty()) {
+            logger.debug("Finished loading meta-data of bundle '{}'.", bundle.getSymbolicName());
+            for (T object : queue.removeAll(bundle.getBundleId())) {
+                action.apply(object);
+            }
+        } else {
+            logger.debug("'{}' still vetoed by '{}', queueing '{}'", bundle.getSymbolicName(),
+                    vetoes.get(bundle.getBundleId()), queue.get(bundle.getBundleId()));
+        }
+    }
+
+    private Bundle getBundle(final Class<?> classFromBundle) {
+        final ClassLoader classLoader = AccessController.doPrivileged(new PrivilegedAction<ClassLoader>() {
+            @Override
+            public ClassLoader run() {
+                return classFromBundle.getClassLoader();
+            }
+        });
+        if (classLoader instanceof BundleReference) {
+            Bundle bundle = ((BundleReference) classLoader).getBundle();
+            logger.trace("Bundle of {} is {}", classFromBundle, bundle.getSymbolicName());
+            return bundle;
+        }
+        return null;
+    }
+
+    /**
+     * Run the given action.
+     *
+     * If necessary, it will get queued and executed once the binding's XML files have been loaded.
+     *
+     * @param object the argument for the action
+     */
+    public void applyActionFor(final T object) {
+        boolean veto = false;
+        Bundle bundle = getBundle(object.getClass());
+        long bundleId = bundle.getBundleId();
+        for (BundleProcessor proc : bundleProcessors) {
+            if (!proc.hasFinishedLoading(bundle)) {
+                veto = true;
+                if (!vetoes.containsEntry(bundleId, proc)) {
+                    logger.trace("Marking '{}' as vetoed by '{}'", bundle.getSymbolicName(), proc);
+                    vetoes.put(bundleId, proc);
+                }
+            } else {
+                logger.trace("'{}' already finished processing '{}'", proc, bundle.getSymbolicName());
+            }
+        }
+        if (veto) {
+            if (!queue.containsEntry(bundle, object)) {
+                logger.trace("Queueing '{}' in bundle '{}'", object, bundle.getSymbolicName());
+                queue.put(bundleId, object);
+            }
+            logger.debug("Meta-data of bundle '{}' is not fully loaded ({}), deferring action for '{}'",
+                    bundle.getSymbolicName(), vetoes.get(bundleId), object);
+        } else {
+            logger.trace("No veto for bundle '{}', directly executing the action", bundle.getSymbolicName());
+            action.apply(object);
+        }
+    }
+
+    /**
+     * Add a {@link BundleProcessor} to listen to.
+     *
+     * @param bundleProcessor
+     */
+    public void addBundleProcessor(BundleProcessor bundleProcessor) {
+        bundleProcessors.add(bundleProcessor);
+        bundleProcessor.registerListener(this);
+    }
+
+    /**
+     * Remove a {@link BundleProcessor} again.
+     *
+     * @param bundleProcessor
+     */
+    public void removeBundleProcessor(BundleProcessor bundleProcessor) {
+        bundleProcessor.unregisterListener(this);
+        bundleProcessors.remove(bundleProcessor);
+    }
+
+}

--- a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java
@@ -26,7 +26,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 
 import org.eclipse.smarthome.config.core.BundleProcessor;
-import org.eclipse.smarthome.config.core.BundleProcessor.BundleProcessorListener;
+import org.eclipse.smarthome.config.core.BundleProcessorVetoManager;
 import org.eclipse.smarthome.config.core.ConfigDescription;
 import org.eclipse.smarthome.config.core.ConfigDescriptionParameter;
 import org.eclipse.smarthome.config.core.ConfigDescriptionRegistry;
@@ -65,7 +65,6 @@ import org.eclipse.smarthome.core.thing.util.ThingHandlerHelper;
 import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.State;
 import org.osgi.framework.Bundle;
-import org.osgi.framework.BundleReference;
 import org.osgi.service.component.ComponentContext;
 import org.osgi.service.event.EventHandler;
 import org.slf4j.Logger;
@@ -95,8 +94,7 @@ import com.google.common.collect.SetMultimap;
  * @author Kai Kreuzer - Removed usage of itemRegistry and thingLinkRegistry, fixed vetoing mechanism
  * @author Andre Fuechsel - Added the {@link ThingTypeMigrationService} 
  */
-public class ThingManager extends AbstractItemEventSubscriber
-        implements ThingTracker, BundleProcessorListener, ThingTypeMigrationService {
+public class ThingManager extends AbstractItemEventSubscriber implements ThingTracker, ThingTypeMigrationService {
 
     private static final String FORCEREMOVE_THREADPOOL_NAME = "forceRemove";
 
@@ -607,74 +605,36 @@ public class ThingManager extends AbstractItemEventSubscriber
         if (!isHandlerRegistered(thing)) {
             return;
         }
-        synchronized (thing) {
-            if (!isInitializing(thing)) {
-                doInitializeHandler(thing, thing.getHandler());
-            } else {
-                logger.warn("Attempt to initialize a handler twice for thing '{}' at the same time will be ignored.",
-                        thing.getUID());
-            }
-        }
+        initializationVetoManager.applyActionFor(thing.getHandler());
     }
 
-    private void doInitializeHandler(Thing thing, ThingHandler thingHandler) {
-        if (!isVetoed(thingHandler)) {
-            ThingType thingType = getThingType(thing);
-            applyDefaultConfiguration(thing, thingType);
-            if (isInitializable(thing, thingType)) {
-                setThingStatus(thing, buildStatusInfo(ThingStatus.INITIALIZING, ThingStatusDetail.NONE));
-                doInitializeHandler(thingHandler);
-            } else {
-                logger.debug("Thing '{}' not initializable, check required configuration parameters.", thing.getUID());
-                setThingStatus(thing,
-                        buildStatusInfo(ThingStatus.UNINITIALIZED, ThingStatusDetail.HANDLER_CONFIGURATION_PENDING));
-            }
-        }
-    }
-
-    private boolean isVetoed(final ThingHandler thingHandler) {
-        boolean veto = false;
-        Bundle bundle = getBundle(thingHandler.getClass());
-        for (BundleProcessor proc : bundleProcessors) {
-            if (!proc.hasFinishedLoading(bundle)) {
-                veto = true;
-                if (!initializerVetoes.containsEntry(bundle, proc)) {
-                    logger.trace("Marking '{}' vetoed by '{}'", bundle.getSymbolicName(), proc);
-                    initializerVetoes.put(bundle, proc);
+    private final BundleProcessorVetoManager<ThingHandler> initializationVetoManager = new BundleProcessorVetoManager<>(
+            new BundleProcessorVetoManager.Action<ThingHandler>() {
+                @Override
+                public void apply(ThingHandler thingHandler) {
+                    Thing thing = thingHandler.getThing();
+                    synchronized (thing) {
+                        if (!isInitializing(thing)) {
+                            ThingType thingType = getThingType(thing);
+                            applyDefaultConfiguration(thing, thingType);
+                            if (isInitializable(thing, thingType)) {
+                                setThingStatus(thing,
+                                        buildStatusInfo(ThingStatus.INITIALIZING, ThingStatusDetail.NONE));
+                                doInitializeHandler(thingHandler);
+                            } else {
+                                logger.debug("Thing '{}' not initializable, check required configuration parameters.",
+                                        thing.getUID());
+                                setThingStatus(thing, buildStatusInfo(ThingStatus.UNINITIALIZED,
+                                        ThingStatusDetail.HANDLER_CONFIGURATION_PENDING));
+                            }
+                        } else {
+                            logger.warn(
+                                    "Attempt to initialize a handler twice for thing '{}' at the same time will be ignored.",
+                                    thing.getUID());
+                        }
+                    }
                 }
-            } else {
-                logger.trace("'{}' already finished processing '{}'", proc, bundle.getSymbolicName());
-            }
-        }
-        if (veto) {
-            if (!initializerQueue.containsEntry(bundle, thingHandler)) {
-                logger.trace("Queueing '{}' in bundle '{}'", thingHandler, bundle.getSymbolicName());
-                initializerQueue.put(bundle.getBundleId(), thingHandler);
-            }
-            logger.debug(
-                    "Meta-data of bundle '{}' is not fully loaded ({}), deferring handler initialization for thing '{}'",
-                    bundle.getSymbolicName(), initializerVetoes.get(bundle), thingHandler.getThing().getUID());
-        } else {
-            logger.debug("Finished loading meta-data of bundle '{}'", bundle.getSymbolicName());
-        }
-        return veto;
-    }
-
-    private Bundle getBundle(final Class<?> classFromBundle) {
-        ClassLoader classLoader = AccessController.doPrivileged(new PrivilegedAction<ClassLoader>() {
-            @Override
-            public ClassLoader run() {
-                return classFromBundle.getClassLoader();
-            }
-        });
-
-        if (classLoader instanceof BundleReference) {
-            Bundle bundle = ((BundleReference) classLoader).getBundle();
-            logger.trace("Bundle of {} is {}", classFromBundle, bundle.getSymbolicName());
-            return bundle;
-        }
-        return null;
-    }
+            });
 
     private void applyDefaultConfiguration(Thing thing, ThingType thingType) {
         if (thingType != null) {
@@ -745,21 +705,6 @@ public class ThingManager extends AbstractItemEventSubscriber
                 }
             }
         }, 0, TimeUnit.NANOSECONDS);
-    }
-
-    @Override
-    public void bundleFinished(BundleProcessor context, Bundle bundle) {
-        initializerVetoes.remove(bundle, context);
-        if (initializerVetoes.get(bundle).isEmpty()) {
-            synchronized (initializerQueue) {
-                for (ThingHandler thingHandler : initializerQueue.removeAll(bundle.getBundleId())) {
-                    initializeHandler(thingHandler.getThing());
-                }
-            }
-        } else {
-            logger.debug("'{}' still vetoed by '{}'", bundle.getSymbolicName(), initializerVetoes.get(bundle));
-            logger.debug("'{}' queued '{}'", bundle.getSymbolicName(), initializerQueue.get(bundle.getBundleId()));
-        }
     }
 
     private boolean isInitializing(Thing thing) {
@@ -1137,14 +1082,11 @@ public class ThingManager extends AbstractItemEventSubscriber
     }
 
     protected void setBundleProcessor(BundleProcessor bundleProcessor) {
-        logger.trace("Added '{}'", bundleProcessor);
-        bundleProcessors.add(bundleProcessor);
-        bundleProcessor.registerListener(this);
+        initializationVetoManager.addBundleProcessor(bundleProcessor);
     }
 
     protected void unsetBundleProcessor(BundleProcessor bundleProcessor) {
-        bundleProcessor.unregisterListener(this);
-        bundleProcessors.remove(bundleProcessor);
+        initializationVetoManager.removeBundleProcessor(bundleProcessor);
     }
 
 }

--- a/bundles/model/org.eclipse.smarthome.model.thing.tests/src/org/eclipse/smarthome/model/thing/tests/GenericThingProviderTest.groovy
+++ b/bundles/model/org.eclipse.smarthome.model.thing.tests/src/org/eclipse/smarthome/model/thing/tests/GenericThingProviderTest.groovy
@@ -328,4 +328,98 @@ class GenericThingProviderTest extends OSGiTest {
 
     }
 
+    @Test
+    void 'assert that things can be embedded within bridges in short notation'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+
+        String model =
+                '''
+            Bridge hue:bridge:myBridge @ "basement" [ ip = "1.2.3.4", username = "123" ] {
+                LCT001 bulb1 [ lightId = "1" ] { Switch : notification }
+            }
+            '''
+        modelRepository.addOrRefreshModel(TESTMODEL_NAME, new ByteArrayInputStream(model.bytes))
+        def List<Thing> actualThings = thingRegistry.getAll()
+
+        assertThat actualThings.size(), is(2)
+
+        assertThat actualThings.find {"hue:bridge:myBridge".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:myBridge:bulb1".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:myBridge:bulb1".equals(it.UID.toString())}.getBridgeUID().toString(), is(equalTo("hue:bridge:myBridge"))
+    }
+
+    @Test
+    void 'assert that things can be embedded within bridges in long notation'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+
+        String model =
+                '''
+            Bridge hue:bridge:myBridge @ "basement" [ ip = "1.2.3.4", username = "123" ] {
+                hue:LCT001:bulb1 [ lightId = "1" ] { Switch : notification }
+            }
+            '''
+        modelRepository.addOrRefreshModel(TESTMODEL_NAME, new ByteArrayInputStream(model.bytes))
+        def List<Thing> actualThings = thingRegistry.getAll()
+
+        assertThat actualThings.size(), is(2)
+
+        assertThat actualThings.find {"hue:bridge:myBridge".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:bulb1".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:bulb1".equals(it.UID.toString())}.getBridgeUID().toString(), is(equalTo("hue:bridge:myBridge"))
+    }
+
+    @Test
+    void 'assert that standalone things can have bridges in long notation'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+
+        String model =
+                '''
+            Bridge hue:bridge:myBridge @ "basement" [ ip = "1.2.3.4", username = "123" ]
+            hue:LCT001:bulb1 (hue:bridge:myBridge) [ lightId = "1" ] { Switch : notification }
+            '''
+        modelRepository.addOrRefreshModel(TESTMODEL_NAME, new ByteArrayInputStream(model.bytes))
+        def List<Thing> actualThings = thingRegistry.getAll()
+
+        assertThat actualThings.size(), is(2)
+
+        assertThat actualThings.find {"hue:bridge:myBridge".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:bulb1".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:bulb1".equals(it.UID.toString())}.getBridgeUID().toString(), is(equalTo("hue:bridge:myBridge"))
+    }
+
+    @Test
+    void 'assert that standalone thing with a bridge works in short notation'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+
+        String model =
+                '''
+        Bridge hue:bridge:myBridge @ "basement" [ ip = "1.2.3.4", username = "123" ]
+        LCT001 bulb1 (hue:bridge:myBridge) [ lightId = "1" ] { Switch : notification }
+        '''
+        modelRepository.addOrRefreshModel(TESTMODEL_NAME, new ByteArrayInputStream(model.bytes))
+        def List<Thing> actualThings = thingRegistry.getAll()
+
+        assertThat actualThings.size(), is(2)
+
+        assertThat actualThings.find {"hue:bridge:myBridge".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:bulb1".equals(it.UID.toString())}, is(notNullValue())
+        assertThat actualThings.find {"hue:LCT001:bulb1".equals(it.UID.toString())}.getBridgeUID().toString(), is(equalTo("hue:bridge:myBridge"))
+    }
+
+    @Test
+    void 'assert that standalone thing without a bridge does not work in short notation'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+
+        String model =
+                '''
+        LCT001 bulb1 [ lightId = "1" ] { Switch : notification }
+        hue:LCT001:bulb2 (hue:bridge:myBridge) [ lightId = "2" ] { Switch : notification }
+        '''
+        modelRepository.addOrRefreshModel(TESTMODEL_NAME, new ByteArrayInputStream(model.bytes))
+        def List<Thing> actualThings = thingRegistry.getAll()
+
+        assertThat actualThings.size(), is(1)
+        assertThat actualThings.find {"hue:LCT001:bulb2".equals(it.UID.toString())}, is(notNullValue())
+    }
+
 }

--- a/bundles/model/org.eclipse.smarthome.model.thing.tests/src/org/eclipse/smarthome/model/thing/tests/GenericThingProviderTest4.groovy
+++ b/bundles/model/org.eclipse.smarthome.model.thing.tests/src/org/eclipse/smarthome/model/thing/tests/GenericThingProviderTest4.groovy
@@ -1,0 +1,174 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.model.thing.tests;
+
+import static org.hamcrest.CoreMatchers.*
+import static org.junit.Assert.*
+import static org.junit.matchers.JUnitMatchers.*
+
+import org.eclipse.smarthome.config.core.BundleProcessor
+import org.eclipse.smarthome.config.core.BundleProcessor.BundleProcessorListener
+import org.eclipse.smarthome.core.thing.Bridge
+import org.eclipse.smarthome.core.thing.Thing
+import org.eclipse.smarthome.core.thing.ThingRegistry
+import org.eclipse.smarthome.core.thing.ThingUID
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory
+import org.eclipse.smarthome.model.core.ModelRepository
+import org.eclipse.smarthome.model.thing.test.hue.TestHueThingHandlerFactoryX
+import org.eclipse.smarthome.model.thing.test.hue.TestHueThingTypeProvider
+import org.eclipse.smarthome.test.OSGiTest
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.osgi.framework.Bundle
+import org.osgi.framework.FrameworkUtil
+import org.osgi.service.component.ComponentContext
+
+
+/**
+ * Test class for the GenericThingProvider.
+ *
+ * It focuses on the XML processing (i.e. ThingType loading) and verifies
+ * that the loading of Things gets delayed until the XML processing is completed.
+ *
+ * @author Simon Kaufmann - initial contribution and API.
+ */
+class GenericThingProviderTest4 extends OSGiTest{
+    private TestHueThingTypeProvider thingTypeProvider
+    private BundleProcessor bundleProcessor
+    private Bundle bundle
+    private ThingHandlerFactory hueThingHandlerFactory
+    private BundleProcessorListener listener
+    private boolean finished
+
+    private final static String TESTMODEL_NAME = "testModelX.things"
+
+    ModelRepository modelRepository;
+    ThingRegistry thingRegistry;
+
+    @Before
+    public void setUp() {
+        thingRegistry = getService ThingRegistry
+        assertThat thingRegistry, is(notNullValue())
+        modelRepository = getService ModelRepository
+        assertThat modelRepository, is(notNullValue())
+        modelRepository.removeModel(TESTMODEL_NAME)
+
+        def componentContextMock = [
+            getBundleContext: {getBundleContext()}
+        ] as ComponentContext
+        hueThingHandlerFactory = new TestHueThingHandlerFactoryX(componentContextMock)
+
+        finished = false;
+        bundle = FrameworkUtil.getBundle(TestHueThingHandlerFactoryX)
+        bundleProcessor = [
+            "hasFinishedLoading": { return finished },
+            "registerListener": {l -> listener = l},
+            "unregisterListener": {l -> listener = null}
+        ] as BundleProcessor
+        registerService(bundleProcessor)
+    }
+
+    @After
+    public void tearDown() {
+        modelRepository.removeModel(TESTMODEL_NAME);
+        if (thingTypeProvider != null) {
+            unregisterService(thingTypeProvider)
+        }
+    }
+
+    private def updateModel() {
+        String model =
+                '''
+            Xhue:Xbridge:myBridge [ XipAddress = "1.2.3.4", XuserName = "123" ]
+            '''
+        modelRepository.addOrRefreshModel(TESTMODEL_NAME, new ByteArrayInputStream(model.bytes))
+    }
+
+    private def finishLoading() {
+        thingTypeProvider = new TestHueThingTypeProvider()
+        registerService(thingTypeProvider)
+
+        finished = true;
+        listener.bundleFinished(bundleProcessor, bundle)
+    }
+
+    private void assertThatAllIsGood() {
+        assertThat thingRegistry.getAll().size(), is(1)
+
+        Thing bridge = thingRegistry.get(new ThingUID("Xhue:Xbridge:myBridge"))
+        assertThat bridge, is(notNullValue())
+        assertThat bridge, is(instanceOf(Bridge))
+    }
+
+    @Test
+    void 'assert that things are created only once the bundle finished loading with update_factory_loaded'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+        updateModel()
+        assertThat thingRegistry.getAll().size(), is(0)
+        registerService(hueThingHandlerFactory, ThingHandlerFactory.class.name)
+        assertThat thingRegistry.getAll().size(), is(0)
+        finishLoading()
+        assertThatAllIsGood()
+    }
+
+    @Test
+    void 'assert that things are created only once the bundle finished loading with factory_update_loaded'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+        registerService(hueThingHandlerFactory, ThingHandlerFactory.class.name)
+        assertThat thingRegistry.getAll().size(), is(0)
+        updateModel()
+        assertThat thingRegistry.getAll().size(), is(0)
+        finishLoading()
+        assertThatAllIsGood()
+    }
+
+    @Test
+    void 'assert that things are created only once the bundle finished loading with loaded_factory_update'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+        finishLoading()
+        assertThat thingRegistry.getAll().size(), is(0)
+        registerService(hueThingHandlerFactory, ThingHandlerFactory.class.name)
+        assertThat thingRegistry.getAll().size(), is(0)
+        updateModel()
+        assertThatAllIsGood()
+    }
+
+    @Test
+    void 'assert that things are created only once the bundle finished loading with loaded_update_factory'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+        finishLoading()
+        assertThat thingRegistry.getAll().size(), is(0)
+        updateModel()
+        assertThat thingRegistry.getAll().size(), is(0)
+        registerService(hueThingHandlerFactory, ThingHandlerFactory.class.name)
+        assertThatAllIsGood()
+    }
+
+    @Test
+    void 'assert that things are created only once the bundle finished loading with factory_loaded_update'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+        registerService(hueThingHandlerFactory, ThingHandlerFactory.class.name)
+        assertThat thingRegistry.getAll().size(), is(0)
+        finishLoading()
+        assertThat thingRegistry.getAll().size(), is(0)
+        updateModel()
+        assertThatAllIsGood()
+    }
+
+    @Test
+    void 'assert that things are created only once the bundle finished loading with update_loaded_factory'() {
+        assertThat thingRegistry.getAll().size(), is(0)
+        updateModel()
+        assertThat thingRegistry.getAll().size(), is(0)
+        finishLoading()
+        assertThat thingRegistry.getAll().size(), is(0)
+        registerService(hueThingHandlerFactory, ThingHandlerFactory.class.name)
+        assertThatAllIsGood()
+    }
+}

--- a/bundles/model/org.eclipse.smarthome.model.thing/OSGI-INF/genericthingprovider.xml
+++ b/bundles/model/org.eclipse.smarthome.model.thing/OSGI-INF/genericthingprovider.xml
@@ -16,5 +16,6 @@
    </service>
    <reference bind="setThingTypeRegistry" cardinality="0..1" interface="org.eclipse.smarthome.core.thing.type.ThingTypeRegistry" name="ThingTypeRegistry" policy="dynamic" unbind="unsetThingTypeRegistry"/>
    <reference bind="addThingHandlerFactory" cardinality="0..n" interface="org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory" name="ThingHandlerFactory" policy="dynamic" unbind="removeThingHandlerFactory"/>
+   <reference bind="addBundleProcessor" cardinality="0..n" interface="org.eclipse.smarthome.config.core.BundleProcessor" name="BundleProcessor" policy="dynamic" unbind="removeBundleProcessor"/>
    <reference bind="setLocaleProvider" cardinality="1..1" interface="org.eclipse.smarthome.core.i18n.LocaleProvider" name="LocaleProvider" policy="static" unbind="unsetLocaleProvider"/>
 </scr:component>

--- a/bundles/model/org.eclipse.smarthome.model.thing/src/org/eclipse/smarthome/model/thing/internal/GenericThingProvider.xtend
+++ b/bundles/model/org.eclipse.smarthome.model.thing/src/org/eclipse/smarthome/model/thing/internal/GenericThingProvider.xtend
@@ -7,10 +7,17 @@
  */
 package org.eclipse.smarthome.model.thing.internal
 
+import java.util.ArrayList
 import java.util.Collection
 import java.util.List
+import java.util.Map
 import java.util.Set
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CopyOnWriteArrayList
+import org.eclipse.smarthome.config.core.BundleProcessor
 import org.eclipse.smarthome.config.core.Configuration
+import org.eclipse.smarthome.core.common.registry.AbstractProvider
+import org.eclipse.smarthome.core.i18n.LocaleProvider
 import org.eclipse.smarthome.core.thing.Bridge
 import org.eclipse.smarthome.core.thing.Channel
 import org.eclipse.smarthome.core.thing.ChannelUID
@@ -18,33 +25,27 @@ import org.eclipse.smarthome.core.thing.Thing
 import org.eclipse.smarthome.core.thing.ThingProvider
 import org.eclipse.smarthome.core.thing.ThingTypeUID
 import org.eclipse.smarthome.core.thing.ThingUID
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory
+import org.eclipse.smarthome.core.thing.binding.builder.BridgeBuilder
 import org.eclipse.smarthome.core.thing.binding.builder.ChannelBuilder
 import org.eclipse.smarthome.core.thing.binding.builder.ThingBuilder
 import org.eclipse.smarthome.core.thing.type.ChannelDefinition
+import org.eclipse.smarthome.core.thing.type.ChannelKind
+import org.eclipse.smarthome.core.thing.type.ChannelTypeUID
 import org.eclipse.smarthome.core.thing.type.ThingTypeRegistry
+import org.eclipse.smarthome.core.thing.type.TypeResolver
+import org.eclipse.smarthome.core.thing.util.ThingHelper
 import org.eclipse.smarthome.model.core.ModelRepository
 import org.eclipse.smarthome.model.core.ModelRepositoryChangeListener
+import org.eclipse.smarthome.model.thing.thing.ModelBridge
 import org.eclipse.smarthome.model.thing.thing.ModelChannel
 import org.eclipse.smarthome.model.thing.thing.ModelPropertyContainer
 import org.eclipse.smarthome.model.thing.thing.ModelThing
 import org.eclipse.smarthome.model.thing.thing.ThingModel
+import org.eclipse.xtend.lib.annotations.Data
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
-import org.eclipse.smarthome.core.thing.binding.builder.BridgeBuilder
-import java.util.Map
-import org.eclipse.smarthome.core.thing.util.ThingHelper
-import java.util.concurrent.ConcurrentHashMap
-import org.eclipse.smarthome.model.thing.thing.ModelBridge
-import org.eclipse.smarthome.core.common.registry.AbstractProvider
-import org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory
-import java.util.concurrent.CopyOnWriteArrayList
-import org.eclipse.smarthome.model.thing.internal.GenericThingProvider.QueueContent
-import java.util.ArrayList
-import org.eclipse.smarthome.core.thing.type.TypeResolver
-import java.util.Locale
-import org.eclipse.smarthome.core.i18n.LocaleProvider
-import org.eclipse.smarthome.core.thing.type.ChannelKind
-import org.eclipse.smarthome.core.thing.type.ChannelTypeUID
+import org.eclipse.smarthome.config.core.BundleProcessorVetoManager
 
 /**
  * {@link ThingProvider} implementation which computes *.things files.
@@ -54,7 +55,8 @@ import org.eclipse.smarthome.core.thing.type.ChannelTypeUID
  *         https://bugs.eclipse.org/bugs/show_bug.cgi?id=450236 - Considering
  *         ThingType Description
  * @author Simon Kaufmann - Added asynchronous retry in case the handler 
- *         factory cannot load a thing yet (bug 470368)
+ *         factory cannot load a thing yet (bug 470368), 
+ *         added delay until ThingTypes are fully loaded
  * @author Markus Rathgeb - Add locale provider support
  * 
  */
@@ -72,7 +74,16 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
 
     private val List<QueueContent> queue = new CopyOnWriteArrayList
     private var Thread lazyRetryThread = null
-
+    
+    private val BundleProcessorVetoManager<ThingHandlerFactory> vetoManager = new BundleProcessorVetoManager(new BundleProcessorVetoManager.Action<ThingHandlerFactory>() {
+        override apply(ThingHandlerFactory thingHandlerFactory) {
+            thingsMap.keySet.forEach [
+                // create things for this specific thingHandlerFactory from the model.
+                createThingsFromModelForThingHandlerFactory(it, thingHandlerFactory)
+            ]
+        }
+    })
+    
     private static final Logger logger = LoggerFactory.getLogger(GenericThingProvider)
 
     def void activate() {
@@ -87,21 +98,53 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
 
     def private void createThingsFromModel(String modelName) {
         logger.debug("Read things from model '{}'", modelName);
-
-        val things = newArrayList
+        if (thingsMap.get(modelName) == null) {
+            thingsMap.put(modelName, newArrayList)
+        }
         if (modelRepository != null) {
             val model = modelRepository.getModel(modelName) as ThingModel
-            if (model != null) {
-                model.things.forEach [
-                    createThing(null, things)
-                ]
+            model?.things.map[
+                // Get the ThingHandlerFactories
+                val ThingUID thingUID = constructThingUID
+                if (thingUID != null) {
+                    val thingTypeUID = constructThingTypeUID(thingUID)
+                    return thingHandlerFactories.findFirst[
+                        supportsThingType(thingTypeUID)
+                    ]
+                } else {
+                    // ignore the Thing because its definition is broken
+                    return null
+                }
+            ].filter[
+                // Drop it if there is no ThingHandlerFactory yet which can handle it 
+                it != null
+            ].toSet.forEach[
+                // Execute for each unique ThingHandlerFactory
+                vetoManager.applyActionFor(it)
+            ]
+        }
+    }
+    
+    def private ThingUID constructThingUID(ModelThing modelThing) {
+        if (modelThing.id != null) {
+            return new ThingUID(modelThing.id)
+        } else {
+            if (modelThing.bridgeUID != null) {
+                val bindingId = new ThingUID(modelThing.bridgeUID).bindingId
+                return new ThingUID(bindingId, modelThing.thingTypeId, modelThing.thingId)
+            } else {
+                logger.warn("Thing {} does not have a bridge so it needs to be defined in full notation like <bindingId>:{}:{}", modelThing.thingTypeId, modelThing.thingTypeId, modelThing.thingId)
+                return null
             }
         }
-        thingsMap.put(modelName, things)
     }
-
-    def private void createThing(ModelThing modelThing, Bridge parentBridge, Collection<Thing> thingList) {
-        createThing(modelThing, parentBridge, thingList, null)
+    
+    def private ThingTypeUID constructThingTypeUID(ModelThing modelThing, ThingUID thingUID) {
+        if (modelThing.thingTypeId != null) {
+            return new ThingTypeUID(thingUID.bindingId, modelThing.thingTypeId)
+        } else {
+            return new ThingTypeUID(thingUID.bindingId, thingUID.thingTypeId)
+        }
     }
 
     def private void createThing(ModelThing modelThing, Bridge parentBridge, Collection<Thing> thingList,
@@ -111,14 +154,21 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
         var ThingUID bridgeUID = null
         if (parentBridge != null) {
             val bindingId = parentBridge.thingTypeUID.bindingId
-            val thingTypeId = modelThing.thingTypeId
-            val thingId = modelThing.thingId
-            thingTypeUID = new ThingTypeUID(bindingId, thingTypeId)
-            thingUID = new ThingUID(thingTypeUID, thingId, parentBridge.parentPath)
+            if (modelThing.id != null) {
+                thingUID = modelThing.constructThingUID
+                thingTypeUID = new ThingTypeUID(bindingId, thingUID.thingTypeId)
+            } else {
+                thingTypeUID = new ThingTypeUID(bindingId, modelThing.thingTypeId)
+                thingUID = new ThingUID(thingTypeUID, modelThing.thingId, parentBridge.parentPath)
+            }
             bridgeUID = parentBridge.UID
         } else {
-            thingUID = new ThingUID(modelThing.id)
-            thingTypeUID = new ThingTypeUID(thingUID.bindingId, thingUID.thingTypeId)
+            thingUID = modelThing.constructThingUID
+            if (thingUID == null) {
+                // ignore the Thing because its definition is broken
+                return
+            }
+            thingTypeUID = modelThing.constructThingTypeUID(thingUID)
             if (modelThing.bridgeUID != null && !modelThing.bridgeUID.empty) {
                 bridgeUID = new ThingUID(modelThing.bridgeUID)
             }
@@ -140,7 +190,7 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
 
         val thingType = thingTypeUID.thingType
 
-        val label = if(modelThing.label != null) modelThing.label else thingType?.label
+        val label = if (modelThing.label != null) modelThing.label else thingType?.label
         
         val location = modelThing.location
 
@@ -352,36 +402,18 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
             switch type {
                 case org.eclipse.smarthome.model.core.EventType.ADDED: {
                     createThingsFromModel(modelName)
-                    val things = thingsMap.get(modelName) ?: newArrayList
-                    things.forEach [
-                        notifyListenersAboutAddedElement
-                    ]
                 }
                 case org.eclipse.smarthome.model.core.EventType.MODIFIED: {
                     val oldThings = thingsMap.get(modelName) ?: newArrayList
-                    val oldThingsUIDs = oldThings.map[UID].toList
-
-                    createThingsFromModel(modelName)
-                    val currentThings = thingsMap.get(modelName) ?: newArrayList
-                    val currentThingUIDs = currentThings.map[UID].toList
-                    val removedThings = oldThings.filter[!currentThingUIDs.contains(UID)]
-                    val addedThings = currentThings.filter[!oldThingsUIDs.contains(UID)]
-
+                    val model = modelRepository.getModel(modelName) as ThingModel
+                    val removedThings = oldThings.filter[!model.things.map[
+                        new ThingUID(it.id)
+                    ].contains(it.UID)]
                     removedThings.forEach [
                         notifyListenersAboutRemovedElement
                     ]
-                    addedThings.forEach [
-                        notifyListenersAboutAddedElement
-                    ]
 
-                    currentThings.forEach [ newThing |
-                        val oldThing = oldThings.findFirst[it.UID == newThing.UID]
-                        if (oldThing != null) {
-                            if (!ThingHelper.equals(oldThing, newThing)) {
-                                notifyListenersAboutUpdatedElement(oldThing, newThing)
-                            }
-                        }
-                    ]
+                    createThingsFromModel(modelName)
                 }
                 case org.eclipse.smarthome.model.core.EventType.REMOVED: {
                     val things = thingsMap.remove(modelName) ?: newArrayList
@@ -425,10 +457,7 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
     }
 
     def private thingHandlerFactoryAdded(ThingHandlerFactory thingHandlerFactory) {
-        thingsMap.keySet.forEach [
-            // create things for this specific thingHandlerFactory from the model.
-            createThingsFromModelForThingHandlerFactory(it, thingHandlerFactory)
-        ]
+        vetoManager.applyActionFor(thingHandlerFactory)
     }
 
     def private createThingsFromModelForThingHandlerFactory(String modelName, ThingHandlerFactory factory) {
@@ -445,9 +474,10 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
         newThings.forEach [ newThing |
             val oldThing = oldThings.findFirst[it.UID == newThing.UID]
             if (oldThing != null) {
-                // this thing already existed, so let's ignore it
+                logger.debug("Updating thing '{}' from model '{}'.", newThing.UID, modelName);
+                notifyListenersAboutUpdatedElement(oldThing, newThing)
             } else {
-                logger.debug("Adding thing '{}' from model '{}.", newThing.UID, modelName);
+                logger.debug("Adding thing '{}' from model '{}'.", newThing.UID, modelName);
                 thingsMap.get(modelName).add(newThing)
                 newThing.notifyListenersAboutAddedElement
             }
@@ -506,6 +536,14 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
         ThingUID thingUID
         ThingUID bridgeUID
         ThingHandlerFactory thingHandlerFactory
+    }
+
+    def protected void addBundleProcessor(BundleProcessor bundleProcessor) {
+        vetoManager.addBundleProcessor(bundleProcessor)
+    }
+
+    def protected void removeBundleProcessor(BundleProcessor bundleProcessor) {
+        vetoManager.removeBundleProcessor(bundleProcessor)
     }
 
 }

